cprogs/readbuffer.1
.TH readbuffer 1 2001-10-21 chiark-backup
.SH NAME
readbuffer \- read input from devices which don't like constant stopping and starting
.SH SYNOPSIS
.B readbuffer
.RB [ --mlock ]
.RI [ size ]
.SH DESCRIPTION
.B readbuffer
reads data on standard input and writes it to standard output.  It
will internally buffer up to \fIsize\fR megabytes of data, and will
only read more data when the buffer is at least 75% empty.
.PP
\fIsize\fR may also be suffixed with
.BR m ", " k ", or " b
to indicate that it is in megabytes (2^20), kilobytes (2^10) or bytes.
.PP
It is intended for use in situations where many small
reads are undesirable for performance reasons, e.g. tape drives.
.SH OPTIONS
.TP
.B --mlock
Calls
.BR mlock (2)
to lock the buffer into memory.
.SH "SEE ALSO"
.BR writebuffer (1),
.BR mlock (2)


debian/changelog
chiark-utils (8.0.0-0deepin1) unstable; urgency=medium

  * feat: add sw64 support 

 -- hudeng <hudeng@deepin.org>  Tue, 21 Oct 2025 17:32:41 +0800

chiark-utils (8.0.0) unstable; urgency=medium

  * Rewrite the packaging to use `debhelper` and `dh` in as standard a way
    as possible.  As a consequence, support root-less builds.
    No functional changes to the installed packages.
    Closes: #1089303.  [Niels Thykier]

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Sun, 12 Jan 2025 20:19:58 +0000

chiark-utils (7.0.3) unstable; urgency=medium

  * Team upload.
  * git-cache-proxy: Handle extra-parameters in git protocol (Closes: #1040476).
    Thanks to Anthony PERARD for the patch.
  * git-branchmove: Always pass --no-verify to 'git push' (Closes: #1080082).

 -- Sean Whitton <spwhitton@spwhitton.name>  Thu, 26 Sep 2024 17:07:41 +0100

chiark-utils (7.0.2) unstable; urgency=medium

  * debian/rules: Pass -D_TIME_BITS=64 too.  Closes: #1078190.
  * Change file size type to unsigned long long.  Closes: #1075821.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Thu, 15 Aug 2024 22:28:46 +0100

chiark-utils (7.0.1) unstable; urgency=medium

  * backup/snaprsync: add -AX to the rsync rune
  * Enable LFS and 64-bit time_t support.  Closes: #1067157.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Thu, 04 Jul 2024 12:53:58 +0100

chiark-utils (7.0.0) unstable; urgency=medium

  New facility:
  * prefork-interp, Proc::Prefork::Interp: new facility

  Improvements:
  * expire-iso8601: Replace with new implementation.  Closes:#862897.

  Bugfixes:
  * acctdump: Fix manual page section.
  * git-cache-proxy: fix for new git.  Closes: #1006695.  [Anthony Perard]
  * git-branchmove: install manpage..  Closes: #946364.  [Sean Whitton]
  * hexterm: Update #! and Suggest tcl8.6.  Closes: #905474.

  Build and packaging improvements:
  * Fix cross build.  Closes: #951242.  [Helmut Grohne]
  * copyright: Be clear that SPDX indications are licence grants
  * Add lintian override for python dependency
  * Add lintian override for chiark-really
  * fix priority "extra" to "optional"
  * Fix build-dependency on debhelper.
  * Set source format to 3.0 (native)
  * Add Vcs-* fields.
  * Update Standards-Version
  * timespeccmp.h: Move use to just cgi-fcgi-interp

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 22 Aug 2022 00:06:57 +0100

chiark-utils (6.1.2+nmu1) unstable; urgency=medium

  * Non-maintainer upload.
    - Maintainer has approved this via personal communication.
  * Bump debhelper compat to 13 (Closes: #965453).

 -- Sean Whitton <spwhitton@spwhitton.name>  Tue, 21 Jul 2020 08:17:49 -0700

chiark-utils (6.1.2) unstable; urgency=medium

  * git-cache-proxy: Recheck gc timestamp after taking lock
  * git-cache-proxy: Tolerate failure to write gc stamp
  * git-cache-proxy: Do not update timestamp due to housekeeping gc

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Wed, 12 Feb 2020 16:10:02 +0000

chiark-utils (6.1.2~citrix2~) unstable; urgency=medium

  * git-cache-proxy: Move $SIG{ALRM} setting to global
  * git-cache-proxy: Abandon fetch attempt if client disconnects
  * git-cache-proxy: Use open-coded fork/exec for git gc

 -- Ian Jackson <ian.jackson@citrix.com>  Tue, 11 Feb 2020 18:34:58 +0000

chiark-utils (6.1.1) unstable; urgency=medium

  * fishdescriptor: Use Python "errno" module
  * git-cache-proxy: Periodically run `git gc --quiet'
  * git-cache-proxy: Document some options in the head comment

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Tue, 11 Feb 2020 16:37:31 +0000

chiark-utils (6.0.5~citrix1) unstable; urgency=medium

  * fishdescriptor: Use Python "errno" module, not "os.errno"
    (which is not in modern python3 apparently).

 -- Ian Jackson <ian.jackson@citrix.com>  Fri, 20 Dec 2019 17:50:02 +0000

chiark-utils (6.1.0~iwj1) unstable; urgency=medium

  * really: Writeability of rc.local suffices, too.
    (Non-sysvinit compatibility.)

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 12 Aug 2019 10:59:02 +0100

chiark-utils (6.1.0) unstable; urgency=medium

  * Non-maintainer upload.
    - Maintainer has approved this via personal communication.
  * git-branchmove: rewrite in perl (Closes: #914398, #914399)
    - Add dependencies on libgit-wrapper-perl, libtry-tiny-perl to
      bin:chiark-scripts.
  * git-branchmove: new --detach feature.

 -- Sean Whitton <spwhitton@spwhitton.name>  Sat, 07 Dec 2019 14:10:26 -0700

chiark-utils (6.0.4) unstable; urgency=medium

  * sync-accounts: Fix perl syntax error.  Closes:#865985.
  * changelog: Document bug number for bugfix in 6.0.4~citrix1.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Fri, 12 Apr 2019 11:48:00 +0100

chiark-utils (6.0.4~citrix1) unstable; urgency=medium

  * fishdescriptor: cast __errno_location correctly.  Closes:#926858.

 -- Ian Jackson <ian.jackson@citrix.com>  Mon, 08 Apr 2019 17:03:47 +0100

chiark-utils (6.0.3) unstable; urgency=medium

  * Upload to Debian unstable.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Tue, 20 Nov 2018 18:25:53 +0000

chiark-utils (6.0.3~citrix1) unstable; urgency=medium

  * git-branchmove: Avoid transporting tags about
  * fishdescriptor: Cope if donor is in a chroot without /proc
  * fishdescriptor: Cope if only donor has /run/user/UID

 -- Ian Jackson <ian.jackson@citrix.com>  Mon, 19 Nov 2018 16:11:55 +0000

chiark-utils (6.0.2) unstable; urgency=medium

  * Rebuild with arch:all packages included in the upload.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 11 Jun 2018 14:35:08 +0100

chiark-utils (6.0.1) unstable; urgency=low

  fishdescriptor:
  * Fix numerous bugs, affecting the `exec' feature with nontrivial fd
    specifications.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Mon, 11 Jun 2018 14:31:24 +0100

chiark-utils (6.0.0) unstable; urgency=medium

  xbatmon-simple:
  * Minor improvements to the .txt doc.

 -- Ian Jackson <ijackson@chiark.greenend.org.uk>  Fri, 20 Apr 2018 16:56:04 +0100


cprogs/readbuffer.c
/*
 * readbuffer.c
 *
 * A program for reading input from devices which don't like constant
 * stopping and starting, such as tape drives.  readbuffer is:
 *  Copyright (C) 1997-1998,2000-2001 Ian Jackson <ian@chiark.greenend.org.uk>
 *
 * readbuffer is part of chiark backup, a system for backing up GNU/Linux and
 * other UN*X-compatible machines, as used on chiark.greenend.org.uk.
 * chiark backup is:
 *  Copyright (C) 1997-1998,2000-2001 Ian Jackson <ian@chiark.greenend.org.uk>
 *  Copyright (C) 1999 Peter Maydell <pmaydell@chiark.greenend.org.uk>
 *
 * This is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 3,
 * or (at your option) any later version.
 *
 * This is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this file; if not, consult the Free Software
 * Foundation's website at www.fsf.org, or the GNU Project website at
 * www.gnu.org.
 *
 */

#include "rwbuffer.h"

const char *progname= "readbuffer";

static size_t waitempty;

int main(int argc, const char *const *argv) {
  int r,reading;

  startup(argv);
  waitempty= (buffersize*1)/4;
  reading=1;
  maxselfd=2;
  
  while (!seeneof || used) {
    
    FD_ZERO(&readfds);
    if (reading) {
      if (used<buffersize-1) {
	FD_SET(0,&readfds);
      } else {
	reading=0;
      }
    }
    FD_ZERO(&writefds); if (used) FD_SET(1,&writefds);

    callselect();

    if (FD_ISSET(0,&readfds)) {
      r= read(0,rp,min(buffersize-1-used,buf+buffersize-rp));
      if (!r) {
        seeneof=1; reading=0;
      } else if (r<0) {
        if (!(errno == EAGAIN || errno == EINTR)) { perror("read"); exit(1); }
      } else {
        used+= r;
        rp+= r;
        if (rp == buf+buffersize) rp=buf;
      }
    }

    if (FD_ISSET(1,&writefds)) {
      assert(used);
      r= write(1,wp,min(used,buf+buffersize-wp));
      if (r<=0) {
        if (!(errno == EAGAIN || errno == EINTR)) { perror("write"); exit(1); }
      } else {
        used-= r;
        wp+= r;
        if (wp == buf+buffersize) wp=buf;
      }
      if (used < waitempty && !seeneof) {
	reading=1;
      }
    }
  }
  exit(0);
}
